### DISCUSSION TITLE
Design problem: Flexible scalar-or-array types

### DISCUSSION THREAD

**Discussion by bilderbuchi** [OpenModelica/OpenModelica#12842] [Q&A]
In a large package, we have implement a class hierarchy. So far in these models we have O(10) scalar quantities that represent one physical element, with appropriate start values and parameter defaults.
Now, we are extending these to a different branch, where these quantities are now present as **arrays**, not a **scalars**, representing that these now have a stack of elements instead of just one.
Most often, the defaults/start values are the same for all elements. 

* We want to retain scalar access as `single.var` in the old/majority of classes, and 
* We don't want to use single-element arrays for the scalar classes as we want to avoid our users having to type/define scalars everywhere as `param={1.4}` just to deal with a design issue
* We want to enable access to the multi class quantities as `multi.var[idx]`
* We would like to preserve DRY and **not duplicate defaults/start values** across both "single" and "multi" variants, as that is prone to miss updates/changes down the line.

To achieve this change in an elegant way within Modelica's type system, we have implemented a `replaceable type RealScalarOrArray` that is either `Real` (the "single" case up to now) or `Real[count]` (in the "multi" case).
This has worked alright so far, but for defining appropriate defaults/start values in one place, we run into the boundaries of Modelica expressiveness. 

I want to get experienced eyes and feedback on this, to see if this could be achieved in a better way, or if this an OM bug or two somewhere. 
Also, I maybe want to suggest a relaxation of the Modelica spec to enable this use case, but I want to get this vetted first before going to the MLS tracker.
@casella 
@perost this is also the background behind #12838, the problem with `fill(1.4)` is new since your fix. I don't know if this a shortcoming of the fix or was just unearthed by it.

It seems to me like the `BaseWorkaround` approach is viable, but I feel like I should not have to do the dance with a custom identity element to paper over possibly-too-strict broadcast/array rules of Modelica. E.g., if assignment would broadcast, `BaseBroadcast` would just work and be simple.

The MWE:
```modelica
package FlexibleTypesWithEach "Enable scalar or array variables and DRY pattern for default values"

  model BaseBroadcast "If broadcasting would work over assignment, we could do this simply"
    replaceable type RealScalarOrArray = Real;
    RealScalarOrArray var(each min = 0, each max = 2) = 1.3; //Error: Non-array modification '1.3' for array component 'var', possibly due to missing 'each'. 
    parameter RealScalarOrArray param(each min = 0, each max = 2) = 1.4; //Non-array modification '1.4' for array component 'param', possibly due to missing 'each'.
  end BaseBroadcast;

  model BaseEach "Let's try with a way to use each"
    replaceable type RealScalarOrArray = Real;
    // Have to use start, as each outside not allowed
    RealScalarOrArray var(each min = 0, each max = 2, each start = 1.3);
    parameter RealScalarOrArray param(each min = 0, each max = 2, each start = 1.4); //No matching function found for fill(1.4). Candidates are: fill(Any, Integer, ...) => Any[:, ...]
    // presumably from the Single single instantiation

    // Also, we get warnings about  Parameter single.param/multi.param has no value, and is fixed during initialization (fixed=true), using available start value (start=1.4) as default value.
    //annoying but expected
  end BaseEach;

  model BaseWorkaround
    replaceable type RealScalarOrArray = Real;
    RealScalarOrArray var(each min = 0, each max = 2) = 1.3*flex_one;
    parameter RealScalarOrArray param(each min = 0, each max = 2) = 1.4*flex_one;

    // Need to define a "dummy" algebraic identity element just to make the dimensions work.
    parameter RealScalarOrArray flex_one = 1.0 "Either scalar or array of ones";
    
    //Also: why does this accept "each" when in Single, where RealScalarOrArray is just Real? See MLS 7.2.5 Modifiers for Array Elements
    // "The each keyword on a modifier requires that it is applied in an array declaration/modification"
  end BaseWorkaround;
  
  model Base 
//  extends BaseBroadcast; // this would be most convenient
//  extends BaseEach; // warning about start as default is inconvenient
    extends BaseWorkaround; // observes DRY, needs an artificial element
  end Base;
  
  model Single
  extends Base;
  end Single;

  model Multi
  extends Base(
    redeclare replaceable type RealScalarOrArray=Real[count], // the flex type is now an array
    flex_one=ones(count) // only in case of BaseWorkaround
  );
  parameter Integer count = 1;
  end Multi;
  
  model MWE
  Single single "Enable access as single.var and single.param";
  Multi multi(count=3) "Enable access as multi.var[2] and multi.param[2]";
  end MWE;
end FlexibleTypesWithEach;
```

I'll be grateful for any feedback!  🙇 🙏

**Comment by perost** [comment]
I don't think you're actually allowed to change the number of dimensions like that. [7.3.2 in the specification ](https://specification.modelica.org/master/inheritance-modification-and-redeclaration.html#constraining-type) states:
>Similarly the type used in a redeclaration must have the same number of dimensions as the type of redeclared element.

We don't seem to be checking for that though, but we probably should. So some of the issues you're seeing might be due to the compiler just ignoring the dimensions in some cases since the number of dimensions is not supposed to change.

**Reply by bilderbuchi** [reply]
Oh no! 🙅 😮 That is worse news than expected!

Do you know of another pattern to achieve what we want?

**Reply by perost** [reply]
Not really. You could implement the scalar case as arrays with one element and have alias variables for them to access the "scalars". But I'm not sure if that really works well in practice since Modelica doesn't allow you to have to variables that represent the same data, only variables that are equal to each other.

But having types which are sometimes scalars and sometimes arrays would be tricky, in Modelica you're usually only allowed to change the sizes of dimensions and not the number of dimensions.

**Reply by bilderbuchi** [reply]
> But having types which are sometimes scalars and sometimes arrays would be tricky, in Modelica you're usually only allowed to change the sizes of dimensions and not the number of dimensions.

Would you judge it to have realistic chances to get this Modelica restriction relaxed somewhat (i.e. only for the `constrainedby` relationship), to accommodate the present use case? I.e. should I bother writing something up for the MLS tracker, or would that be wasted effort because it would be rejected immediately? @casella ?

Would "[empty arrays](https://specification.modelica.org/master/arrays.html#empty-arrays)" be an option somehow (dimension size zero feels somewhat equivalent to a scalar), or are those only suitable for dimension elimination etc, without useful values? 

Maybe related: https://github.com/modelica/ModelicaSpecification/issues/721 https://github.com/modelica/ModelicaSpecification/issues/1247

**Comment by bilderbuchi** [comment]
Maybe I could use `break` to remove and recreate `var` and `param`? 🤔  It would have to become  a models, blocks or connectors, though, which complicates matters somewhat, probably to the point of infeasibility.

**Comment by casella** [comment]
@bilderbuchi, I would suggest to post the question and the MWE on the [Modelica Specification issue tracker](https://github.com/modelica/ModelicaSpecification/issues/new).

**Reply by bilderbuchi** [reply]
Thanks for your judgement, as I said above I wanted to wait for something like this before going further:

> Would you judge it to have realistic chances to get this Modelica restriction relaxed somewhat (i.e. only for the constrainedby relationship), to accommodate the present use case? I.e. should I bother writing something up for the MLS tracker, or would that be wasted effort because it would be rejected immediately? @casella ?
